sympy/matrices/determinant.py

```diff
@@ -648,15 +648,14 @@ def _det(M, method="bareiss", iszerofunc=None):
     n = M.rows
 
     if n == M.cols: # square check is done in individual method functions
+        # 1. if M is triangular - calculate the product of the diagonal entries
         if M.is_upper or M.is_lower:
             m = 1
             for i in range(n):
                 m = m * M[i, i]
             return _get_intermediate_simp(_dotprodsimp)(m)
-        elif n == 0:
-            return M.one
-        elif n == 1:
-            return M[0,0]
+        # 2. if M is 2x2 or 3x3 - let's use the hardcoded formula
+        #    (not checking 0x0 and 1x1 matrices as they are _always_ upper & lower triangular)
```

Comment:
In the `_det()` method, `n == 0` and `n == 1` conditions weren't covered by tests because they never get executed - both the 1x1 and 0x0 matrices are `M.is_upper` and `M.is_lower`, so the 'multiply the diagonal' method covers them.
We could move these conditions up, however I decided to go for removing them altogether - diagonal multiplication is about as fast.

Reply:
I think a better solution is to test for `n == 0` and `n == 1` first, although that of course depends on how often this case happens compared to the others.

However, the comments are a really nice addition and I think that the current version is also OK. However, there are some merge conflicts.

Reply:
I agree, especially now that @sylee957 switched the **triangularity** check to the **strongly connected components** check (https://github.com/sympy/sympy/pull/21474).
I will check if any of my tests in this PR are still relevant.

Reply:
Checked, the tests are still relevant. This can be merged to increase the coverage.